src/main/c/Makefile

```diff
@@ -1,17 +1,95 @@
-.PHONY: all test clean libsimdpp
+# This makefile calls itself recursively to build the libraries with
+# different compiler options for different ABI versions:
+#   -fabi-version=2 was the default from g++-3.4 to g++-4.9
+#   -fabi-version=9 was the default in g++-5.x
+
+# On MacOS, libc++ is the default library, and the default clang-based
+# compiler doesn't accept the -fabi-version=N option.  We still use the 
+# name "libhail_abi_v9.dylib" to mean "libhail compatible with recent abi
+
+ifndef SUBMAKE_NAME
+
+.PHONY: top all clean debug prebuilt test
+
+top: all
+
+CXX_VERSION :=$(shell $(CXX) --version | head -1)
+ifneq ($(filter clang llvm LLVM,$(CXX_VERSION)),)
+  CXX_TYPE :=clang
+else
+  CXX_TYPE :=not_clang
+endif
+
+CXX_ABI_V9_ERRS :=$(shell echo "" >a.cpp; $(CXX) -fabi-version=9 -c a.cpp 2>a.err; rm -f a.cpp a.o 2>/dev/null; cat a.err)
+
+
+
+all clean debug prebuilt test:
+ifeq ($(CXX_ABI_V9_ERRS),)
+	@$(MAKE) --no-print-directory SUBMAKE_NAME=abi_new SUBMAKE_CXXFLAGS=-fabi-version=9 $@
+else
+	@$(MAKE) --no-print-directory SUBMAKE_NAME=abi_new SUBMAKE_CXXFLAGS= $@
+endif
+
+else
+
+UNAME_S :=$(shell uname -s)
+UNAME_P :=$(shell uname -p)
+
+# Resolve the full pathnames of executables so that any unintentional
+# use of non-system executables will be visible in the build log.
+
+CAT :=$(shell which cat 2>/dev/null)
+CAT ?=cat
```

Comment:
This doesn't work:

```
$ cat Makefile
foo := $(shell which foo)
foo ?= foo
all:
	echo $(foo)
$ which foo
$ make
echo 

```

Reply:
You're right, empty string is not equivalent to undefined.  Will fix.